File: macos/Packages/MilmilKit/Sources/MilmilAPI/Endpoints/AuthEndpoints.swift
import Foundation

// Mirrors web/src/hooks/use-auth.ts and web/src/lib/api/setup.ts.
public extension APIClient {
    func health() async throws -> Health {
        try await get("/health")
    }

    func setupStatus() async throws -> SetupStatus {
        try await get("/api/v1/setup/status")
    }

    func authStatus() async throws -> AuthStatus {
        try await get("/api/v1/auth/status")
    }

    /// `device_name` is what shows up in the token list; pass
    /// `DeviceName.current()` so it reads "milmil for macOS — <hostname>".
    func login(username: String, password: String, deviceName: String?) async throws -> LoginOutcome {
        try await post("/api/v1/auth/login", body: LoginRequest(username: username, password: password, deviceName: deviceName))
    }

    func completeTwoFactor(userID: String, code: String, deviceName: String?) async throws -> LoginSession {
        try await post("/api/v1/auth/login/2fa", body: TwoFactorRequest(userID: userID, code: code, deviceName: deviceName))
    }

    func me() async throws -> User {
        try await get("/api/v1/auth/me")
    }

    // MARK: Avatar

    /// Upload an image (png / jpeg / webp, ≤ 2 MB after the client's downscale).
    func uploadAvatar(_ data: Data, filename: String, mimeType: String) async throws -> AvatarResponse {
        try await upload("/api/v1/auth/me/avatar", fileField: "file", data: data, filename: filename, mimeType: mimeType)
    }

    /// Have the server copy an image it can reach (a character portrait).
    func setAvatar(sourceURL: URL) async throws -> AvatarResponse {
        try await put("/api/v1/auth/me/avatar", body: AvatarSourceRequest(sourceURL: sourceURL.absoluteString))
    }

    func deleteAvatar() async throws {
        try await delete("/api/v1/auth/me/avatar")
    }

    /// `avatar_url` is server-relative; `size` picks the 128 / 512 variant.
    nonisolated func avatarURL(_ path: String?, size: Int? = nil) -> URL? {
        guard let path, !path.isEmpty, var components = URLComponents(string: path) else { return nil }
        if let size {
            var items = components.queryItems ?? []
            items.removeAll { $0.name == "size" }
            items.append(URLQueryItem(name: "size", value: String(size)))
            components.queryItems = items
        }
        return components.url(relativeTo: baseURL)?.absoluteURL
    }

    /// Deletes the current token row server-side. Clear the local copy after.
    func logout() async throws {
        try await post("/api/v1/auth/logout")
    }

    func apiTokens() async throws -> [APIToken] {
        try await get("/api/v1/api-tokens")
    }

    func revokeAPIToken(id: String) async throws {
        try await delete("/api/v1/api-tokens/\(id)")
    }

    func webSocketTicket() async throws -> WebSocketTicket {
        try await get("/api/v1/ws/ticket")
    }
}

public enum DeviceName {
    /// "milmil for macOS — Pie" (truncated server-side to 100 chars).
    public static func current() -> String {
        let host = Host.current().localizedName ?? ProcessInfo.processInfo.hostName
        return "milmil for macOS — \(host)"
    }
}

File: android/core/api/src/main/kotlin/dev/milmil/api/AuthEndpoints.kt
package dev.milmil.api

import io.ktor.http.HttpMethod

/** Mirrors `MilmilKit/Endpoints/AuthEndpoints.swift` and `web/src/lib/api/auth.ts`. */
public suspend fun ApiClient.health(): Health =
    MilmilJson.decodeFromString(Health.serializer(), execute(HttpMethod.Get, "/health", null))

public suspend fun ApiClient.setupStatus(): SetupStatus =
    MilmilJson.decodeFromString(SetupStatus.serializer(), execute(HttpMethod.Get, "/api/v1/auth/status", null))

public suspend fun ApiClient.me(): User =
    MilmilJson.decodeFromString(User.serializer(), execute(HttpMethod.Get, "/api/v1/auth/me", null))

/**
 * Rate limited to 0.2 req/s per IP with burst 10 — a wrong password must not be
 * retried in a loop; [ApiError.RateLimited] carries the server's Retry-After.
 */
public suspend fun ApiClient.login(username: String, password: String, deviceName: String? = null): LoginOutcome {
    val payload = MilmilJson.encodeToString(LoginRequest.serializer(), LoginRequest(username, password, deviceName))
    val raw = execute(HttpMethod.Post, "/api/v1/auth/login", payload)
    val response = MilmilJson.decodeFromString(LoginResponse.serializer(), raw)
    val userId = response.userId
    return when {
        response.token != null && response.user != null ->
            LoginOutcome.Session(LoginSession(response.token, response.user))
        response.twoFactorRequired && userId != null -> LoginOutcome.TwoFactorRequired(userId)
        else -> throw ApiError.Status(200, raw)
    }
}

public suspend fun ApiClient.completeTwoFactor(userId: String, code: String, deviceName: String? = null): LoginSession {
    val payload = MilmilJson.encodeToString(TwoFactorRequest.serializer(), TwoFactorRequest(userId, code, deviceName))
    val raw = execute(HttpMethod.Post, "/api/v1/auth/2fa/verify", payload)
    val response = MilmilJson.decodeFromString(LoginResponse.serializer(), raw)
    val token = response.token
    val user = response.user
    if (token == null || user == null) throw ApiError.Status(200, raw)
    return LoginSession(token, user)
}

public suspend fun ApiClient.logout() {
    execute(HttpMethod.Post, "/api/v1/auth/logout", null)
}
